<a href="https://colab.research.google.com/github/Rustam64/AI/blob/main/Week6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [9]:
import numpy as np
import tensorflow as tf
import matplotlib.pyplot as plt

In [10]:
#Load Data
mnist = tf.keras.datasets.mnist

#Split data into train and test
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

#Normalize images
train_images=train_images/255.0
test_images=test_images/255.0

In [11]:
#2.1 Advanced Activity: Implement a CNN with Dropout and Batch Normalization
def evaluate_cnn():
    model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(128, activation='relu'),
        tf.keras.layers.BatchNormalization(),  # Batch Normalization layer
        tf.keras.layers.Dropout(0.5),  # Dropout layer with a dropout rate of 0.5
        tf.keras.layers.Dense(10, activation='softmax')
    ])

    model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=10)
    test_loss, test_acc = model.evaluate(test_images, test_labels)
    print(f"Test accuracy with CNN, Dropout, and Batch Normalization: {test_acc}")

    return model

cnn_model = evaluate_cnn()


# Dropout is a regularization technique used during training to prevent overfitting.
# It works by randomly setting a fraction of the input units to a layer to zero during each forward and backward pass.
# During training, at each iteration, dropout randomly "drops out" (sets to zero) a fraction of the neurons in the network.

# Batch Normalization is used to improve the stability and convergence of neural networks during training by normalizing the inputs to each layer.
# Batch Normalization computes the mean and standard deviation of the input data within a mini-batch during training.
# It then scales and shifts the inputs to ensure that they have a specified mean and standard deviation (typically close to 0 and 1).
# This normalization is applied before the activation function of each layer.

Epoch 1/10
1875/1875 [==============================] - 24s 12ms/step - loss: 0.1763 - accuracy: 0.9474
Epoch 2/10
1875/1875 [==============================] - 23s 12ms/step - loss: 0.0884 - accuracy: 0.9737
Epoch 3/10
1875/1875 [==============================] - 23s 12ms/step - loss: 0.0679 - accuracy: 0.9790
Epoch 4/10
1875/1875 [==============================] - 23s 12ms/step - loss: 0.0529 - accuracy: 0.9838
Epoch 5/10
1875/1875 [==============================] - 23s 12ms/step - loss: 0.0477 - accuracy: 0.9850
Epoch 6/10
1875/1875 [==============================] - 23s 12ms/step - loss: 0.0437 - accuracy: 0.9861
Epoch 7/10
1875/1875 [==============================] - 23s 12ms/step - loss: 0.0367 - accuracy: 0.9882
Epoch 8/10
1875/1875 [==============================] - 23s 12ms/step - loss: 0.0339 - accuracy: 0.9893
Epoch 9/10
1875/1875 [==============================] - 22s 12ms/step - loss: 0.0298 - accuracy: 0.9898
Epoch 10/10
313/313 [==============================] - 2s 5ms/st

In [12]:
# 2.2 Activity 2: Experiment with Different Activation Functions
def evaluate_cnn_with_activation(activation_function):
    model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(128, activation=activation_function),  # Experiment with different activations here
        tf.keras.layers.Dense(10, activation='softmax')
    ])

    model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=10)
    test_loss, test_acc = model.evaluate(test_images, test_labels)
    print(f"Test accuracy with {activation_function} activation: {test_acc}")

    return model

for activation_function in ['relu', 'sigmoid', 'tanh', 'elu']:
    print(f"Experimenting with {activation_function} activation:")
    evaluate_cnn_with_activation(activation_function)

# Relu had the lowest accuracy, followed by sigmoid while elu had the highest followed by tahn.

Experimenting with relu activation:
Epoch 1/10
1875/1875 [==============================] - 23s 12ms/step - loss: 0.1482 - accuracy: 0.9560
Epoch 2/10
1875/1875 [==============================] - 22s 12ms/step - loss: 0.0512 - accuracy: 0.9841
Epoch 3/10
1875/1875 [==============================] - 27s 14ms/step - loss: 0.0317 - accuracy: 0.9902
Epoch 4/10
1875/1875 [==============================] - 22s 12ms/step - loss: 0.0213 - accuracy: 0.9932
Epoch 5/10
1875/1875 [==============================] - 22s 12ms/step - loss: 0.0139 - accuracy: 0.9955
Epoch 6/10
1875/1875 [==============================] - 22s 12ms/step - loss: 0.0103 - accuracy: 0.9965
Epoch 7/10
1875/1875 [==============================] - 22s 12ms/step - loss: 0.0077 - accuracy: 0.9975
Epoch 8/10
1875/1875 [==============================] - 23s 12ms/step - loss: 0.0064 - accuracy: 0.9976
Epoch 9/10
1875/1875 [==============================] - 22s 12ms/step - loss: 0.0047 - accuracy: 0.9986
Epoch 10/10
313/313 [=======

In [16]:
# 2.3 Activity 3: Add More Convolutional Layers
def evaluate_cnn_with_layer(activation_function):
    model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
        tf.keras.layers.MaxPooling2D((2, 2)),

        tf.keras.layers.Conv2D(64, (3, 3), activation='relu'),  # Additional Conv2D layer
        tf.keras.layers.MaxPooling2D((2, 2)),

        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(128, activation=activation_function),  # Experiment with different activations here
        tf.keras.layers.Dense(10, activation='softmax')
    ])

    model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=10)
    test_loss, test_acc = model.evaluate(test_images, test_labels)
    print(f"Test accuracy: {test_acc}")

    return model

cnn_model = evaluate_cnn_with_layer('relu')

Epoch 1/10
1875/1875 [==============================] - 32s 17ms/step - loss: 0.1331 - accuracy: 0.9596
Epoch 2/10
1875/1875 [==============================] - 30s 16ms/step - loss: 0.0442 - accuracy: 0.9863
Epoch 3/10
1875/1875 [==============================] - 30s 16ms/step - loss: 0.0307 - accuracy: 0.9902
Epoch 4/10
1875/1875 [==============================] - 30s 16ms/step - loss: 0.0223 - accuracy: 0.9933
Epoch 5/10
1875/1875 [==============================] - 30s 16ms/step - loss: 0.0160 - accuracy: 0.9949
Epoch 6/10
1875/1875 [==============================] - 30s 16ms/step - loss: 0.0125 - accuracy: 0.9958
Epoch 7/10
1875/1875 [==============================] - 30s 16ms/step - loss: 0.0098 - accuracy: 0.9967
Epoch 8/10
1875/1875 [==============================] - 31s 16ms/step - loss: 0.0078 - accuracy: 0.9973
Epoch 9/10
1875/1875 [==============================] - 31s 17ms/step - loss: 0.0065 - accuracy: 0.9979
Epoch 10/10
313/313 [==============================] - 2s 5ms/st

In [17]:
# 2.4 Activity 4: Experiment with Different Optimizers
def evaluate_cnn_with_layer(activation_function):
    model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(128, activation=activation_function),  # Experiment with different activations here
        tf.keras.layers.Dense(10, activation='softmax')
    ])

    model.compile(optimizer='SGD', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=10)
    test_loss, test_acc = model.evaluate(test_images, test_labels)
    print(f"Test accuracy: {test_acc}")

    return model

cnn_model = evaluate_cnn_with_layer('relu')

Epoch 1/10
1875/1875 [==============================] - 20s 11ms/step - loss: 0.4582 - accuracy: 0.8753
Epoch 2/10
1875/1875 [==============================] - 20s 11ms/step - loss: 0.2172 - accuracy: 0.9362
Epoch 3/10
1875/1875 [==============================] - 20s 11ms/step - loss: 0.1661 - accuracy: 0.9504
Epoch 4/10
1875/1875 [==============================] - 21s 11ms/step - loss: 0.1356 - accuracy: 0.9596
Epoch 5/10
1875/1875 [==============================] - 23s 12ms/step - loss: 0.1151 - accuracy: 0.9648
Epoch 6/10
1875/1875 [==============================] - 20s 11ms/step - loss: 0.1006 - accuracy: 0.9695
Epoch 7/10
1875/1875 [==============================] - 20s 11ms/step - loss: 0.0900 - accuracy: 0.9729
Epoch 8/10
1875/1875 [==============================] - 19s 10ms/step - loss: 0.0816 - accuracy: 0.9757
Epoch 9/10
1875/1875 [==============================] - 21s 11ms/step - loss: 0.0740 - accuracy: 0.9775
Epoch 10/10
313/313 [==============================] - 1s 4ms/st

In [18]:
# 2.4 Activity 4: Experiment with Different Optimizers
def evaluate_cnn_with_layer(activation_function):
    model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(128, activation=activation_function),  # Experiment with different activations here
        tf.keras.layers.Dense(10, activation='softmax')
    ])

    model.compile(optimizer='RMSprop', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=10)
    test_loss, test_acc = model.evaluate(test_images, test_labels)
    print(f"Test accuracy: {test_acc}")

    return model

cnn_model = evaluate_cnn_with_layer('relu')

Epoch 1/10
1875/1875 [==============================] - 22s 12ms/step - loss: 0.1657 - accuracy: 0.9493
Epoch 2/10
1875/1875 [==============================] - 21s 11ms/step - loss: 0.0540 - accuracy: 0.9838
Epoch 3/10
1875/1875 [==============================] - 21s 11ms/step - loss: 0.0356 - accuracy: 0.9893
Epoch 4/10
1875/1875 [==============================] - 21s 11ms/step - loss: 0.0258 - accuracy: 0.9923
Epoch 5/10
1875/1875 [==============================] - 22s 12ms/step - loss: 0.0191 - accuracy: 0.9942
Epoch 6/10
1875/1875 [==============================] - 22s 12ms/step - loss: 0.0145 - accuracy: 0.9960
Epoch 7/10
1875/1875 [==============================] - 21s 11ms/step - loss: 0.0101 - accuracy: 0.9972
Epoch 8/10
1875/1875 [==============================] - 21s 11ms/step - loss: 0.0076 - accuracy: 0.9979
Epoch 9/10
1875/1875 [==============================] - 21s 11ms/step - loss: 0.0048 - accuracy: 0.9987
Epoch 10/10
313/313 [==============================] - 1s 4ms/st

In [19]:
# 2.4 Activity 4: Experiment with Different Optimizers
def evaluate_cnn_with_layer(activation_function):
    model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(128, activation=activation_function),  # Experiment with different activations here
        tf.keras.layers.Dense(10, activation='softmax')
    ])

    model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=10)
    test_loss, test_acc = model.evaluate(test_images, test_labels)
    print(f"Test accuracy: {test_acc}")

    return model

cnn_model = evaluate_cnn_with_layer('relu')

Epoch 1/10
1875/1875 [==============================] - 22s 12ms/step - loss: 0.1494 - accuracy: 0.9554
Epoch 2/10
1875/1875 [==============================] - 22s 12ms/step - loss: 0.0526 - accuracy: 0.9837
Epoch 3/10
1875/1875 [==============================] - 22s 12ms/step - loss: 0.0342 - accuracy: 0.9892
Epoch 4/10
1875/1875 [==============================] - 23s 12ms/step - loss: 0.0229 - accuracy: 0.9930
Epoch 5/10
1875/1875 [==============================] - 21s 11ms/step - loss: 0.0164 - accuracy: 0.9948
Epoch 6/10
1875/1875 [==============================] - 22s 12ms/step - loss: 0.0123 - accuracy: 0.9962
Epoch 7/10
1875/1875 [==============================] - 22s 12ms/step - loss: 0.0083 - accuracy: 0.9974
Epoch 8/10
1875/1875 [==============================] - 22s 12ms/step - loss: 0.0072 - accuracy: 0.9973
Epoch 9/10
1875/1875 [==============================] - 23s 12ms/step - loss: 0.0058 - accuracy: 0.9980
Epoch 10/10
313/313 [==============================] - 1s 4ms/st

### **RMSprop had the best performance at 0.985 accuracy, followed by adam at 0.9825 and lastly, relu at 0.976.**